### PR TITLE
[CASObjectFormats/NestedV1] Simplify implementation of `decomposeAndSortEdges`

### PR DESCRIPTION
**Pull request by akyrtzi**
Uses the passed-in `SmallVector`s directly and only one additional `SmallVector` and `SmallDenseMap`.
Also avoids the need to use `jitlink::Edge *` as a `DenseMap` key, which helps with making the code more generic later on.

### CODE REVIEW

File: llvm/lib/CASObjectFormats/NestedV1.cpp [hunk 1]
```diff
@@ -624,7 +624,7 @@ static bool compareSymbolsBySemanticsAnd(
     function_ref<bool(const jitlink::Symbol *, const jitlink::Symbol *)>
         NextCompare) {
   if (LHS == RHS)
-    return false;
+    return NextCompare(LHS, RHS);
```
Comment (dexonsmith) [inline llvm/lib/CASObjectFormats/NestedV1.cpp:627]: Haven't had a chance to think through the rest yet, but can you explain why this is necessary? If the symbols are equal, why compare further? (Can/should it be split to a different commit since it seems like a functionality change?)

Comment (dexonsmith) [reply llvm/lib/CASObjectFormats/NestedV1.cpp:627]: Ah, I see the call in `visitEdgesInStableOrder()`, I think I understand.

File: llvm/lib/CASObjectFormats/NestedV1.cpp [hunk 2]
```diff
@@ -748,138 +779,75 @@ static Error decomposeAndSortEdges(
...
-    Edges.push_back(&E);
-    const jitlink::Symbol &S = E.getTarget();
+  SmallVector<EdgeTarget, 16> EdgeTargets;
+  Fixups.reserve(Block.edges_size());
+  TIs.reserve(Block.edges_size());
+
+  visitEdgesInStableOrder(Block, [&](const jitlink::Edge &E) {
+    Fixups.push_back(Fixup{E.getKind(), E.getOffset()});
+    // We'll fill the real index later, using '~0U' to mark as 'unset'.
+    TIs.push_back(TargetInfo{E.getAddend(), ~0U});
```
Comment (dexonsmith) [inline llvm/lib/CASObjectFormats/NestedV1.cpp]: Can you use a name for the `~0U` both here and below?

e.g.,
```
constexpr unsigned AbstractBackedgeIndexPlaceholder = ~0U`;
```

Up to you what to name it... or if you really prefer the magic constant in both places I can live with that.

**Review by dexonsmith — APPROVED** [review]
This is a big cleanup, thanks! One nitpick inline.

File: llvm/lib/CASObjectFormats/NestedV1.cpp (hunk 1 above)
Comment (akyrtzi) [reply llvm/lib/CASObjectFormats/NestedV1.cpp:627]: > Ah, I see the call in visitEdgesInStableOrder()

Right, I also checked the other uses of that function and it doesn't change existing functionality, AFAICT.

File: llvm/lib/CASObjectFormats/NestedV1.cpp (hunk 2 above)
Comment (akyrtzi) [reply llvm/lib/CASObjectFormats/NestedV1.cpp]: Good idea!
